megablocks/layers/mlp.py

```diff
@@ -376,24 +376,152 @@ def forward(self, x, topo):
         return stk.ops.dsd(gelu.gelu(x), w2)
 
 
+class MemoryOptimizedGroupedMLP(torch.autograd.Function):
+    """GroupedMLP with manually scheduled memory reuse."""
+
+    @staticmethod
+    @torch.cuda.amp.custom_fwd
+    def forward(ctx, x, w1, w2, batch_sizes, num_input_bits, num_remat_bits):
+        # x: [m, k], w1: [n, k], w2: [n, k]
+        if (not x.is_contiguous() or not w1.is_contiguous() or
+            not w2.is_contiguous()):
+            raise ValueError("Expected contiguous 'x', 'w1' and 'w2'.")
+
+        # Layer 0: x @ w1.t().
+        sdd_out = gg.backend.gmm(x, w1, batch_sizes, trans_b=True)
```

Comment:
Why is this called `sdd_out`? Just holdover name from the sparse matmul?

Reply:
Yup, I want to re-factor this to re-use code but I did a copy-paste to get this in ASAP.